src/core/hle/service/hid/hidbus/ringcon.h

```diff
@@ -0,0 +1,302 @@
+};
+static_assert(sizeof(JoyDisableSixAxisPollingEntry) == 0x38,
+              "JoyDisableSixAxisPollingEntry is an invalid size");
+
+struct JoyEnableSixAxisPollingEntry {
+    u64 sampling_number;
+    JoyEnableSixAxisPollingData polling_data;
+};
+static_assert(sizeof(JoyEnableSixAxisPollingEntry) == 0x20,
+              "JoyEnableSixAxisPollingEntry is an invalid size");
+
+struct JoyButtonOnlyPollingEntry {
+    u64 sampling_number;
+    JoyButtonOnlyPollingData polling_data;
+};
+static_assert(sizeof(JoyButtonOnlyPollingEntry) == 0x40,
+              "JoyButtonOnlyPollingEntry is an invalid size");
+
+struct JoyDisableSixAxisDataAccessor {
+    DataAccessorHeader header{};
+    std::array<JoyDisableSixAxisPollingEntry, 0xb> entries{};
+};
+static_assert(sizeof(JoyDisableSixAxisDataAccessor) == 0x298,
+              "JoyDisableSixAxisDataAccessor is an invalid size");
+
+struct JoyEnableSixAxisDataAccessor {
+    DataAccessorHeader header{};
+    std::array<JoyEnableSixAxisPollingEntry, 0xb> entries{};
+};
+static_assert(sizeof(JoyEnableSixAxisDataAccessor) == 0x190,
+              "JoyEnableSixAxisDataAccessor is an invalid size");
+
+struct ButtonOnlyPollingDataAccessor {
+    DataAccessorHeader header;
+    std::array<JoyButtonOnlyPollingEntry, 0xb> entries;
+};
+static_assert(sizeof(ButtonOnlyPollingDataAccessor) == 0x2F0,
+              "ButtonOnlyPollingDataAccessor is an invalid size");
+
+class RingController {
+public:
+    enum class DataValid : u32_le {
+        Valid,
+        BadCRC,
+        Cal,
+    };
+
+    struct FirmwareVersion {
+        u8 sub;
+        u8 main;
+    };
+
+    struct FactoryCalibration {
+        s32_le os_max;
+        s32_le hk_max;
+        s32_le zero_min;
+        s32_le zero_max;
+    };
+
+    struct UserCalibration {
+        s16_le os_max;
+        u16 os_max_crc;
+        s16_le hk_max;
+        u16 hk_crc;
+        s16_le zero;
+        u16 zero_crc;
+    };
+
+    explicit RingController();
+    ~RingController();
+
+    // Updates ringcon transfer memory
+    void Update();
+
+    // Returns the device ID of the joycon
+    u8 GetDeviceId() const;
+
+    // Assigns a command from data
+    bool SetCommand(const std::vector<u8>& data);
+
+    // Enables/disables the device
+    void Enable(bool enable);
+
+    // returns true if ringcon is enabled
+    bool IsEnabled() const;
+
+    // returns true if polling mode enabled
+    bool IsPollingMode() const;
+
+    // returns polling mode
+    JoyPollingMode GetPollingMode() const;
+
+    // Sets and enables JoyPollingMode
+    void SetPollingMode(JoyPollingMode mode);
+
+    // Disables JoyPollingMode
+    void DisablePollingMode();
+
+    // Returns a reply from a command
+    std::vector<u8> GetReply();
+
+    // Loads controller input from settings
+    void OnLoadInputDevices();
+
+    // Called on EnableJoyPollingReceiveMode
+    void SetTransferMemoryPointer(u8* t_mem);
+
+private:
+    enum class RingConCommands : u32_le {
```

Comment:
```suggestion
    enum class RingConCommands : u32 {
```

Otherwise this won't work in cases the `u32_le` is a struct